### PR TITLE
 Vertex AI init commit

### PR DESCRIPTION
**Pull request by Demontego**
This is the first phase of implementing google vertex ai in dspy. Please take a look. Write where and how best to correct, what else should be added.

### CODE REVIEW

**Comment by skucherlapati** [comment]
Would it be cleaner to merge this change with existing [dsp.modules.google](https://github.com/stanfordnlp/dspy/blob/main/dsp/modules/google.py)?

**Comment by Demontego** [comment]
Google [GenAI](https://ai.google.dev/) and Google [Vertex AI](https://cloud.google.com/vertex-ai) are two different frameworks. I think we should keep them in different modules. The issue https://github.com/stanfordnlp/dspy/issues/463 asked to make a separate file googlevertexai.

**Comment by insop** [comment]
@Demontego 
Please also consider adding documentation.

**Comment by Demontego** [comment]
Can you merge this PR?

**Comment by arnavsinghvi11** [comment]
Thanks @Demontego !

@okhat will this need an update to dspy-ai before merging? Also, curious if this library installation can be handled outside of setup.py and mentioning the instructions in the documentation @Demontego ?

File: dsp/modules/googlevertexai.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+from pydantic_core import PydanticCustomError
+
+from dsp.modules.lm import LM
+
+try:
+    import vertexai  # type: ignore[import-untyped]
+    from vertexai.language_models import CodeGenerationModel, TextGenerationModel
+    from vertexai.preview.generative_models import GenerativeModel
+except ImportError:
+    print("Not loading VertexAI because it is not installed.")
```
Comment (arnavsinghvi11) [inline dsp/modules/googlevertexai.py:14]: let's remove the print statement here to avoid printing every time someone imports dspy

**Comment by marshmellow77** [comment]
For what its worth: I have been testing and using this integration already for some projects and it works fine for me.

File: dsp/modules/googlevertexai.py (hunk 1 above)
Comment (Demontego) [reply dsp/modules/googlevertexai.py]: i do it

**Comment by Demontego** [comment]
up

**Comment by arnavsinghvi11** [comment]
Hi @Demontego I missed this before, but could you move the existing docs to the [documentation folder as with the other LMs](https://github.com/stanfordnlp/dspy/tree/main/docs/docs/deep-dive/language_model_clients/remote_models). Will be good to merge after that!

File: dsp/modules/googlevertexai.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+    def __init__(
+        self, model_name: str = "text-bison@002", **kwargs,
+    ):
+        """
+        Parameters
+        ----------
+        model : str
+            Which pre-trained model from Google to use?
+            Choices are [`text-bison@002`]
```
Comment (jderry-lucem) [inline dsp/modules/googlevertexai.py:44]: can you clarify what the supported models are? Additionally, `model` is not a parameter. `model_name` is. This was confusing when I was using this code for a small demo project.

File: dsp/modules/googlevertexai.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+            self.client = model_cls(model_name=model_name, safety_settings=kwargs.get('safety_settings')) # pylint: disable=unexpected-keyword-arg,no-value-for-parameter
+        else:
+            self.client = model_cls.from_pretrained(model_name)
+        self.provider = "googlevertexai"
+        self.kwargs = {
+            **self.kwargs,
+            "temperature": 0.7,
+            "max_output_tokens": 1024,
+            "top_p": 1.0,
+            "top_k": 1,
```
Comment (jderry-lucem) [inline dsp/modules/googlevertexai.py:96]: why are these kwargs available for non-gemini models if they are set to defaults for non-gemini models?

**Comment by RajeshThallam** [comment]
@Demontego Can you please fix the typo on the title of this PR from `Vetrex AI` to `Vertex AI`?

**Review by Josephrp — COMMENTED** [review]
this will open up possibilities for image RAG in dspy , in fact evals on image RAG in DSPy.

i'm currently initializing vertex outside DSPy for this reason.

deploying and finetuning using vertex is a large field to open up to optimizations using DSPY.

**Comment by arnavsinghvi11** [comment]
> Hi @Demontego I missed this before, but could you move the existing docs to the [documentation folder as with the other LMs](https://github.com/stanfordnlp/dspy/tree/main/docs/api/language_model_clients). Will be good to merge after that!

Hi @Demontego , just following up on this. PR is good to merge following this!

**Comment by arnavsinghvi11** [comment]
Thanks @Demontego !

**Pull request merged**